### PR TITLE
Fix distinct issue

### PR DESCRIPTION
**Pull request by Gimly**
We have an issue if two points are considered equal by the equality comparer but don't have the same hashcode, the Distinct fails and doesn't correctly consider the points equal, which result in the two close points being regrouped. This is especially important when the user wants to consider points that are "close" within a certain tolerance.

To fix the issue, we have changed the algorithm to create the same vertex for points that the user would consider equal (using the `Equals` method.

### CODE REVIEW

File: src/Ethereality.DoublyConnectedEdgeList/Dcel.cs [hunk 1]
```diff
@@ -23,18 +23,19 @@ public class Dcel<TEdge, TPoint>
                 faces?.ToList() ?? throw new ArgumentNullException(nameof(faces));
         }
 
-        public IHalfEdge<TEdge, TPoint>? FindHalfEdge(TPoint pointA, TPoint pointB)
-        {
-            return HalfEdges.SingleOrDefault(
+        public IVertex<TEdge, TPoint>? FindVertex(TPoint point) =>
```
Comment (Admir-H) [inline src/Ethereality.DoublyConnectedEdgeList/Dcel.cs]: on ne peut pas le valider (le paramètre) ?

File: src/Ethereality.DoublyConnectedEdgeList/DcelFactory.cs [hunk 2]
```diff
@@ -16,26 +16,41 @@ public class DcelFactory<TEdge, TPoint>
 
         public Dcel<TEdge, TPoint> FromShape(IEnumerable<TEdge> edges)
         {
-            var verticesDictionary = CreateVertices(edges);
+            var enumeratedEdges = edges?.ToList() ?? throw new ArgumentNullException(nameof(edges));
 
-            var halfEdges = CreateHalfEdges(edges, verticesDictionary);
-            List<Face<TEdge, TPoint>> faces = CreateFaces(halfEdges);
+            if (enumeratedEdges.Count == 0)
```
Comment (Admir-H) [inline src/Ethereality.DoublyConnectedEdgeList/DcelFactory.cs:21]: L'opérateur `Any()`, en interne utilise la propriété `Cout`si la source est une `ICollection`. Ca permetterait d'éviter le `var enumeratedEdges = edges?.ToList() ...`

Dans `CreateVertices` tu as un autre `ToList()` et dans `CreateHalfEdges `tu fais un `foreach`.

File: src/Ethereality.DoublyConnectedEdgeList/DcelFactory.cs [hunk 3]
```diff
@@ -16,26 +16,41 @@ public class DcelFactory<TEdge, TPoint>
...
 
-        private static Dictionary<TPoint, Vertex<TEdge, TPoint>> CreateVertices(IEnumerable<TEdge> edges) =>
-            edges.Select(s => s.PointA)
-                 .Concat(edges.Select(s => s.PointB))
-                 .Distinct()
-                 .Select(point => new Vertex<TEdge, TPoint>(point))
-                 .ToDictionary(value => value.OriginalPoint, value => value);
+        private static Dictionary<TPoint, Vertex<TEdge, TPoint>> CreateVertices(IReadOnlyList<TEdge> edges)
+        {
+            var allPoints = edges.SelectMany(edge => new[] {edge.PointA, edge.PointB}).ToList();
```
Comment (Admir-H) [inline src/Ethereality.DoublyConnectedEdgeList/DcelFactory.cs]:
```suggestion
            var allPoints = edges.SelectMany(edge => new[] {edge.PointA, edge.PointB});
```

File: tests/Ethereality.DoublyConnectedEdgeList.Tests/TestPoint.cs [hunk 4]
```diff
@@ -4,27 +4,33 @@ namespace Ethereality.DoublyConnectedEdgeList.Tests
 {
     internal class TestPoint : IEquatable<TestPoint>
```
Comment (Admir-H) [inline tests/Ethereality.DoublyConnectedEdgeList.Tests/TestPoint.cs]:
```suggestion
    internal sealed class TestPoint : IEquatable<TestPoint>
```

A cause du `IEquatable`.

**Review by Admir-H — APPROVED** [review]

File: src/Ethereality.DoublyConnectedEdgeList/Dcel.cs (hunk 1 above)
Comment (Gimly) [reply src/Ethereality.DoublyConnectedEdgeList/Dcel.cs]: done

File: src/Ethereality.DoublyConnectedEdgeList/DcelFactory.cs (hunk 3 above)
Comment (Gimly) [reply src/Ethereality.DoublyConnectedEdgeList/DcelFactory.cs]: done

File: tests/Ethereality.DoublyConnectedEdgeList.Tests/TestPoint.cs (hunk 4 above)
Comment (Gimly) [reply tests/Ethereality.DoublyConnectedEdgeList.Tests/TestPoint.cs]: done

**Review by thomashilke — APPROVED** [review]
Ca me parait bien!

**Review by Admir-H — APPROVED** [review]

**Review by Admir-H — APPROVED** [review]

**Review by Admir-H — APPROVED** [review]

**Pull request merged**